schema/pgsql/schema.sql

```diff
@@ -48,55 +48,84 @@ CREATE TABLE channel (
     -- for now type determines the implementation, in the future, this will need a reference to a concrete
     -- implementation to allow multiple implementations of a sms channel for example, probably even user-provided ones
 
+    changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    deleted boolenum NOT NULL DEFAULT 'n',
+
     CONSTRAINT pk_channel PRIMARY KEY (id)
 );
 
+CREATE INDEX idx_channel_changed_at ON channel(changed_at);
+
 CREATE TABLE contact (
     id bigserial,
     full_name citext NOT NULL,
-    username citext, -- reference to web user
+    username citext, -- reference web user
     default_channel_id bigint NOT NULL REFERENCES channel(id),
 
+    changed_at bigint NOT NULL DEFAULT EXTRACT(EPOCH FROM NOW()) * 1000,
+    deleted boolenum NOT NULL DEFAULT 'n',
+
     CONSTRAINT pk_contact PRIMARY KEY (id),
-    UNIQUE (username)
+    UNIQUE (username) -- column must be NULLed for deletion via "deleted = 'y'"
 );
 
+CREATE INDEX idx_contact_changed_at ON contact(changed_at);
+
 CREATE TABLE contact_address (
     id bigserial,
     contact_id bigint NOT NULL REFERENCES contact(id),
     type text NOT NULL, -- 'phone', 'email', ...
     address text NOT NULL, -- phone number, email address, ...
 
-    CONSTRAINT pk_contact_address PRIMARY KEY (id),
-    UNIQUE (contact_id, type) -- constraint may be relaxed in the future to support multiple addresses per type
```

Comment:
And dropping this constraint also looks new. Why is it removed?

Reply:
This is also based on a discussion with @sukhwinder33445.

The uniqueness constraint forbids re-adding another contact address of a previously deleted - via `deleted = 'y'` - type. When setting the `type` column to `NULL`, as did elsewhere, this would mess up history mentioning deleted entries. As there was still no consensus if we want to support showing deleted entries, the goal was not to purge history.

There was also the idea of having a conditional `UNIQUE` constraint, similar to your discussion in <https://github.com/Icinga/icinga-notifications/pull/191#discussion_r1660678059>.

After going back and forth, we decided to just drop the `UNIQUE` constraint. However, I am very open for other suggestions.

Reply:
> When setting the `type` column to `NULL`, as did elsewhere, this would mess up history mentioning deleted entries.

Does the history currently show the addresses? Should it even do that (that might even bringt up privacy concerns, "John Doe was notified via SMS" and "John Doe was notified by SMS to 0118999881[...]" are quite different things)?

Reply:
Data protection and privacy is even another issue. For example, to comply with the GDPR, it must generally be possible to remove personal data, even if Art. 17 GDPR allows some exceptions. However, law is definitely not my domain.

Back on the technical ground, I would argue in favor of lifting this constraint. It was already marked with a comment to be potentially removed and I don't see any current limitation in its favor. Next to enabling the `deleted = 'y'` logic required here, it would allow, i.e., multiple phone numbers per contact. To stay with your example, some contact might get notified both on their old and new phone number during transition periods, e.g., getting messages both to the old 999 and the new 0118 999 881 999 119 7253.